### PR TITLE
FIX-#2181: constructing OmniSci frame from unsupported types

### LINKED ISSUE #2181: Do fallback to pandas on unsupported dtypes at OmniSci
OmniSci frame currently does not support some data types (for example #1972), for now, OmniSci frame just throwing an exception when we're trying to construct from unsupported types, we need to do fallback in that case.

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2181 
- [x] tests added and passing

### CODE REVIEW

File: modin/experimental/engines/omnisci_on_ray/frame/data.py [hunk 1]
```diff
@@ -67,6 +68,7 @@ def __init__(
         index_cols=None,
         uses_rowid=False,
         force_execution_mode=None,
+        is_default_frame=False,
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/frame/data.py:71]: I think `is_default_frame` doesn't really reflect the meaning of the attribute. Probably `has_unsupported_data`?

File: modin/experimental/engines/omnisci_on_ray/frame/data.py [hunk 2]
```diff
@@ -1212,6 +1224,37 @@ def from_pandas(cls, df):
...
+
+        new_columns = pd.Index(data=at.column_names, dtype="O")
+        new_index = pd.RangeIndex(at.num_rows)
+        new_dtypes = pd.Series(
+            [cls._arrow_type_to_dtype(col.type) for col in at.columns],
+            index=new_columns,
+        )
+
+        if len(unsupported_cols) > 0:
+            ErrorMessage.single_warning(
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/frame/data.py:1245]: I'm not sure we get the desired number of messages here. Consider we have a frame with unsupported data in columns `a` and `b`. Then `df["c"] = df["a"]` would produce three warning messages (for `['a', 'b']`, `['a']`, `['a', 'b', 'c']`). Is it intentional?

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py [hunk 3]
```diff
@@ -47,17 +48,95 @@ def _compute_num_partitions(cls):
         return 1
 
     @classmethod
-    def from_arrow(cls, at, return_dims=False):
+    def from_pandas(cls, df, return_dims=False, unsupported_cols=None):
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py:51]: What is the purpose of the `unsupported_cols` parameter? I don't see a scenario where it is really used by a caller.

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py [hunk 4]
```diff
@@ -47,17 +48,95 @@ def _compute_num_partitions(cls):
...
+                unsupported column types.
+
+        Returns
+        -------
+            Tuple of pyarrow.Table representation of `obj` (for future using) and
+            a list of unsupported columns.
+        """
+        if isinstance(obj, (pandas.Series, pandas.DataFrame)):
+
+            def fast_select_dtypes(df, dtype, nrows=None):
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py:104]: This function selects data, not dtypes, so let's call method appropriately. E.g. `get_object_data_sample`.

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py [hunk 5]
```diff
@@ -47,17 +48,95 @@ def _compute_num_partitions(cls):
...
+                cols = [i for i, col in enumerate(df.dtypes.items()) if col[1] == dtype]
+                if nrows is None:
+                    nrows = len(df)
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py:107]: This would cause out-of-bounds access. Actually, I don't see why we need this parameter, we are not going to call this function with different values. We can simply always access the first row.

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py [hunk 6]
```diff
@@ -47,17 +48,95 @@ def _compute_num_partitions(cls):
...
+                return df.iloc[nrows, cols]
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py:108]: `df.iloc[0][cols]` would allow you to avoid `enumerate` in `cols` selector. IMHO it would look simpler.

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py [hunk 7]
```diff
@@ -47,17 +48,95 @@ def _compute_num_partitions(cls):
...
+
+            # picking first rows from cols with `dtype="object"` to check its actual type,
+            # in case of homogen columns that saves us unnecessary convertion to arrow table
+            type_samples = fast_select_dtypes(obj, dtype="object", nrows=0)
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py:112]: BTW why did you move a part of code into the nested function? This could be two simple rows of code.
```
object_cols = [name for name, dtype in df.dtypes.items() if dtype == "object"]
data_sample = df.iloc[0][object_cols]
```

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py [hunk 8]
```diff
@@ -47,17 +48,95 @@ def _compute_num_partitions(cls):
...
+                return None, unsupported_cols
+
+            try:
+                obj = pyarrow.Table.from_pandas(obj)
+            except pyarrow.lib.ArrowTypeError as e:
+                regex = r"Conversion failed for column ([^\W]*)"
+                unsupported_cols = []
+                for msg in e.args:
+                    match = re.findall(regex, msg)
+                    unsupported_cols.extend(match)
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py:127]: What would happen in case of a mismatched message?

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py [hunk 9]
```diff
@@ -1397,5 +1398,53 @@ def sort(df, cols, ascending, na_position, **kwargs):
...
+    @pytest.mark.parametrize("data", [bad_for_arrow, bad_for_omnisci])
+    def test_construct(self, data):
+        def applier(df, *args, **kwargs):
+            return repr(df)
+
+        run_and_compare(applier, data=data, force_lazy=False)
+
+    def test_from_arrow(self):
+        at = self._get_pyarrow_table(self.bad_for_omnisci)
+        md_df, pd_df = pd.utils.from_arrow(at), pandas.DataFrame(self.bad_for_omnisci)
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:1428]: Two assignments would be easier to read.

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py [hunk 10]
```diff
@@ -1397,5 +1398,53 @@ def sort(df, cols, ascending, na_position, **kwargs):
...
+        # force materialization
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:1429]: We don't need to force materialization, comparison function does it already.

File: modin/experimental/backends/omnisci/query_compiler.py [hunk 11]
```diff
@@ -23,6 +24,59 @@
 from pandas.core.dtypes.common import is_list_like
 
 
+def is_default_frame(value):
```
Comment (ienkovich) [inline modin/experimental/backends/omnisci/query_compiler.py:27]: I'd suggest `inoperable` instead of `default` for naming.

File: modin/experimental/engines/omnisci_on_ray/frame/data.py [hunk 12]
```diff
@@ -1203,7 +1206,16 @@ def from_pandas(cls, df):
             orig_df.index.names = orig_index_names
         new_dtypes = df.dtypes
         df = df.add_prefix("F_")
-        new_parts, new_lengths, new_widths = cls._frame_mgr_cls.from_pandas(df, True)
+
+        partitions, unsupported_cols = cls._frame_mgr_cls.from_pandas(df, True)
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/frame/data.py:1210]: Can we add one more value to the returned tuple instead of packing tuple into another tuple?

**Review by ienkovich — CHANGES_REQUESTED** [review]
Overall the patch looks good! Added functionality is exactly what we need. I have a bunch of minor comments though.

File: modin/experimental/backends/omnisci/query_compiler.py [hunk 13]
```diff
@@ -23,6 +24,59 @@
...
+    if isinstance(value, (tuple, list)):
+        result = False
+        for val in value:
+            result = result or is_default_frame(val)
+        return result
+    elif isinstance(value, dict):
+        return is_default_frame(list(value.values()))
+    else:
+        value = getattr(value, "_query_compiler", value)
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:36]: we're passing API level objects to backends when we're doing groupby on a list of Series whose parent is not grouping frame
```python
df = pd.DataFrame({"a": np.arange(3)})

s1, s2 = pd.Series(np.arange(3)), pd.Series(np.arange(3, 6))
df.groupby([s1, s2])
```

In that case at `df.groupby(...)` we will hit [this](https://github.com/modin-project/modin/blob/dd691cd0d4bb5403b38242563ed8b6ddb316734a/modin/pandas/dataframe.py#L403-L420) if-else branch and will pass pure "by" with API-level Series to GroupByObject. 

GroupByObject constructor [do not](https://github.com/modin-project/modin/blob/dd691cd0d4bb5403b38242563ed8b6ddb316734a/modin/pandas/groupby.py#L43-L92) convert high level objects to query compilers, and `_wrap_aggregation` method also [don't do](https://github.com/modin-project/modin/blob/dd691cd0d4bb5403b38242563ed8b6ddb316734a/modin/pandas/groupby.py#L786-L795) that. 

So in groupby methods of query compiler, we can get API level objects as an arguments

File: modin/experimental/backends/omnisci/query_compiler.py [hunk 14]
```diff
@@ -23,6 +24,59 @@
 from pandas.core.dtypes.common import is_list_like
 
 
+def is_default_frame(value):
+    if isinstance(value, (tuple, list)):
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py:28]: It's a common case when we're getting `np.array` (which is `abc.Iterable`) that represents the whole column of the frame as an argument for the query compiler method (for example `insert` or `groupby_agg`). If it happens, then we'll need to walk through the whole huge column and apply `is_default_frame` to every value in the array. On the other hand, there is a `list` and `tuple` that's often (I think that always) uses to pass lists of query compilers

File: modin/experimental/backends/omnisci/query_compiler.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ienkovich) [reply modin/experimental/backends/omnisci/query_compiler.py:36]: You're right. `_wrap_aggregation` uses API objects to default to pandas, but we should probably extract query compilers otherwise and avoid API objects in a backend (not in this PR though).

File: modin/experimental/backends/omnisci/query_compiler.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ienkovich) [reply modin/experimental/backends/omnisci/query_compiler.py:28]: Good point. Let's keep it as is then.

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py (hunk 8 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py:127]: Yeah, in that case, we'll wrongly assume that there is no unsupported columns, because `len(unsupported_columns) == 0`, even though we've got `ArrowTypeError`. 

We can simply add a check after that for loop if `len(unsupported_columns) == 0` and if it is, then do `unsupported_columns = obj.columns`.

That will make warning about unsupported dtypes inaccurate, because there will be listed all columns of the frame, but if we can't get names of bad columns from arrow error, then we can't get it from anywhere.

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py (hunk 10 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:1429]: yes, the comparison function materializes frames, but it does not trigger any computation to execute.

If we'll create a new modin frame and right after that do `._to_pandas()` then we'll just return our pandas frame from partition and will not hit the execute stage
https://github.com/modin-project/modin/blob/dd691cd0d4bb5403b38242563ed8b6ddb316734a/modin/experimental/engines/omnisci_on_ray/frame/data.py#L905-L907

File: modin/experimental/engines/omnisci_on_ray/frame/data.py (hunk 12 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py]: done

File: modin/experimental/backends/omnisci/query_compiler.py (hunk 11 above)
Comment (dchigarev) [reply modin/experimental/backends/omnisci/query_compiler.py]: done

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py (hunk 9 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py]: done

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py (hunk 7 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py]: yeah, that makes sense, fixed

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py (hunk 6 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py]: thanks for the advice, done

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py (hunk 5 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py]: done

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py (hunk 4 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py]: removed that function

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py (hunk 3 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py]: that was added to be consistent with `from_arrow`, to where I've added that parameter to be able to skip check for unsupported cols, in `from_pandas` we don't use it, so I've removed that parameter

File: modin/experimental/engines/omnisci_on_ray/frame/data.py (hunk 1 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py]: renamed

File: modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py (hunk 8 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py:125]: done that in the last commit

File: modin/experimental/engines/omnisci_on_ray/frame/data.py (hunk 2 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py:1251]: You right about the number of warnings in that case, but all of that warnings will be printed when we're constructing a new frame, and since we want to warn user about unsupported columns at the construction stage, it's not clear how we can decide from modin frame constructor when we should print a warning and when we shouldn't because we have chaining operations.

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py (hunk 10 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:1433]: If we have frames materialized and execution not triggered then we just have nothing to execute. `FrameNode` is always a computation tree list.

Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:1433]: yeah, but what we want to do is trigger the execution stage after constructing frame to check if we build it correctly

File: modin/experimental/engines/omnisci_on_ray/frame/data.py (hunk 2 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py:1251]: Now I think I was wrong about the number of messages. If we warn only in `from_pandas` and `from_arrow` then we would get warnings for leaf frames only and wouldn't get them for other frames in our operations chain.

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py (hunk 10 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:1433]: I'm not sure what you mean by triggering the execution stage. We cannot trigger any execution for `FrameNode`. There is just nothing to execute.

**Comment by ienkovich** [comment]
LGTM! But there is a wrong issue number in commit 3

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py (hunk 10 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:1433]: if we do
```python
pd.DataFrame(unsupported_data)._to_pandas()
```
then we'll just put pandas DataFrame into partition and right after that getting it from there, so execution stage was not triggered and that code will work even at the current master 

if we do
```python
df = pd.DataFrame(unsupported_data)
repr(df)
df._to_pandas()
```
then `repr` will trigger index accessing, frame slicing etc... , so some computation is required and we will potentially hit execution on OmniSci (which we don't want to in case of unsupported data)

**Comment by dchigarev** [comment]
> But there is a wrong issue number in commit 3
thanks for pointing this out)
fixed that

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py (hunk 10 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:1433]: I see. `repr` triggers some slicing but we execute it on an arrow table using `pyarrow`. If the frame was created using `from_pandas` then `repr` might trigger execution on OmniSci (BTW we should use `pyarrow` option here as well since we convert `pandas.DataFrame` to `pyarrow.Table` anyway)

File: modin/experimental/engines/omnisci_on_ray/frame/data.py (hunk 2 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py:1251]: yeah, but if we're doing all operations default to pandas, then at every query compiler function we would do 
1. `.to_pandas()`
2. `func(df)`
3. `from_pandas(res)`

so in your example we will get 3 warnings

Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/frame/data.py:1251]: So I was wrong that I was wrong :)

**Review by ienkovich — APPROVED** [review]
